creusot/src/backend/ty.rs

```diff
@@ -514,12 +514,21 @@ pub(crate) fn destructor<'tcx>(
         ),
     };
 
-    let fail = Expr::Assert(Box::new(Exp::mk_false()), Box::new(Expr::Any));
+    let fail =
+        Expr::BlackBox(Box::new(Expr::Assert(Box::new(Exp::mk_false()), Box::new(Expr::Any))));
+
+    let fields: Vec<_> = fields.iter().cloned().collect();
+    let negative_assertion = if fields.is_empty() {
+        cons_test.neq(Exp::var("input"))
+    } else {
+        Exp::Forall(fields, vec![], Box::new(cons_test.neq(Exp::var("input"))))
+    };
+
```

Comment:
@paulpatault can you confirm that the semantics of this encoding are correct? we want to match the designated constructor and reject other cases.